File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int pt1=height[0];
        int[] left = new int[height.length];
        int[] right = new int[height.length];
        int pt2=height[height.length-1];
        for(int i=0;i<height.length;i++){
            if(height[i]>pt1){
                left[i]=height[i];
                pt1=height[i];
            }else{
                left[i]=pt1;
            }
        }
        for(int j=height.length-1;j>=0;j--){
            if(height[j]>pt2){
                right[j]=height[j];
                pt2=height[j];
            }else{
                right[j]=pt2;
            }
        }
        int ans=0;
        for(int i=0;i<height.length;i++){
                if(left[i]>right[i]){
                    ans=ans+right[i]-height[i];
                }else{
                    ans=ans+left[i]-height[i];
                }
        }
        return ans;
    }
}

File: 0042-trapping-rain-water/0042-trapping-rain-water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function (height) {
    const n = height.length;
    let left = [];
    let right = [];
    left[0] = height[0]
    right[n - 1] = height[n - 1]
    for (let i = 1; i < n; i++) {
        if (height[i] > left[i - 1]) {
            left[i] = height[i]
        } else {
            left[i] = left[i - 1]
        }
    }

    for (let i = n - 2; i >= 0; i--) {
        if (height[i] > right[i + 1]) {
            right[i] = height[i]
        } else {
            right[i] = right[i + 1]
        }
    }

    let ans = 0;

    for (let i = 0; i < n; i++) {
        ans += Math.min(left[i], right[i]) - height[i]
    }
    return ans;
};